docs/rfcs/20210111_centralized_configuration_management.md

```diff
@@ -0,0 +1,588 @@
+    uint16_t port;
+    ss::sstring address;
+};
+```
+
+can be expressed as `<host>:<port>` string, or split the aggregate into separate
+primitive data types.
+
+One of the candidates that may benefit from splitting is the `tls_config`
+
+```c++
+struct key_cert {
+    ss::sstring key_file;
+    ss::sstring cert_file;
+};
+
+class tls_config {
+public:
+   ...
+private:
+    bool _enabled{false};
+    std::optional<key_cert> _key_cert;
+    std::optional<ss::sstring> _truststore_file;
+    bool _require_client_auth{false};
+};
+```
+
+For the `tls_config` it is much more convenient to split it into separate
+independent properties f.e:
+
+`rpc_tls_key`, `rpc_tls_cert`, `rcp_tls_truststore`, ...
+
+#### Environment variables
+
+All command line flags values can be set via environment variables.
+Name of environment variables are always capitalized and words separated with
+underscore character. All environment variables are prefixed with `REDPANDA_`
+prefix. Command line arguments passed to redpanda takes precedence over
+environment variables if both are set.
+
+Example configuration flags and corresponding environment variables:
+
+`--data-directory` `REDPANDA_DATA_DIRECTORY`
+`--heartbeat-interval-ms` `REDPANDA_HEARTBEAT_INTERVAL_MS`
+
+#### Collections
+
+For command line variables collections of parameters is set by passing
+in the same flag more than once.
+f.e:
+
+```bash
+redpanda --seed-server 127.0.0.1:1234 --seed-server 127.0.0.1:1234
+```
+
+This approach is impossible to use with environment variables as each
+assignment overrides previous value. To express collections 
+
+## Configuration property value
+
+Redpanda supports multiple sources for values of configuration properties.
+The configuration property precedence policy clearly defines what would be the
+final value of a configuration property if its value is different in each source.
+Redpanda defines a simple policy for configuration properties value precedence
+as property value is always equal to the one persisted in configuration storage.
+Configuration storage works according to the 'Last write wins' rule.
+This way, the most recent value set is used.
+When redpanda is started with set of flags or environment variables value of
+properties set with flags are persisted and stored values are overridden.
+If configuration change will be propagated to running redpanda instance later on
+it will override stored configuration value.
+This way redpanda always treat stored configuration as a source of truth.
+
+### Features of configuration properties
+
+Configuration properties may have different features.
+Redpanda configuration properties may be described by set of features.
+Those features determine policy of updates, persistance and programmatic API of
+the configuration properties.
+
+#### Scope
+
+Feature describing property range and the at the same time persistence policy.
+
+Possible values: `node/cluster`
+
+- `node` -  properties concern only single node
```

Comment:
Would it make sense to say that any property that affects a single node should be a command line or environmental variable and only global configuration should be set saved in the raft-0 topic?

Reply:
I was considering this, if we will decide that each time `node` property needs to be overrode we have to pass the command line flag then we do not have to store `node` properties other than invariant ones.